README.md
# cerlc

## A Custom Configurable Cyclic Redundancy Check (CRC), Function Generator
 
[![Erlang CI](https://github.com/mdsebald/cerlc/workflows/Erlang%20CI/badge.svg)](https://github.com/mdsebald/cerlc/actions)

### Implemented in native Erlang. No NIFs or depenencies required.

#### Select from over 50 predefined CRC algorithms.  
    See src/cerlc.erl for complete list

#### OR Generate custom CRC algorithms by manually specifying the parameters. 
    bit width, polynomial, initial value, final XOR value, and data reflected or not 

### Build
```
    $ rebar3 compile
```

### Test
```
    $ rebar3 eunit
```

### Use

#### In an Erlang application
```erlang

% Add to list of dependencies in rebar.config
{deps, [
    {cerlc, "0.2.0"}
]}.


% Example: Using a predefined CRC algorithm
CrcDefn = cerlc:init(crc16_aug_ccitt),

% Data may be a binary or list of bytes
Crc = cerlc:calc_crc(Data, CrcDefn)


% Example: Creating a custom CRC algorithm
% Custom CRC parameters: {Bits, Polynomial, InitValue, FinalXorValue, Reflected}
CustomDefn = cerlc:init({8, 16#4F, 0, 16#FF, false}),

Crc = cerlc:calc_crc(Data, CustomDefn)


% Example: Using a macro to evaluate init() function at compile time
-define(CRC_DEFN, cerlc:init(crc16_aug_ccitt)).

Crc = cerlc:calc_crc(Data, ?CRC_DEFN),


```

#### In an Elixir application
```elixir

# Add to list of dependencies in mix.exs
def deps do
  [
    {:cerlc, "~> 0.2.0"}
  ]
end


# Example: Using a predefined CRC algorithm
crc8_defn = :cerlc.init(:crc8)

# Data may be a binary or list of bytes
crc = :cerlc.calc_crc(data, crc8_defn)


# Example: Creating a custom CRC algorithm
# Custom CRC parameters: {bits, polynomial, init_value, final_xor_value, reflected}
custom_defn = cerlc:init({16, 0x1234, 0xFFFF, 0xFFFF, true}),

crc = :cerlc.calc_crc(data, custom_defn)


# Example: Using a module attribute to evaluate init() function at compile time
@crc32_defn :cerlc.init(:crc32_c)

crc = :cerlc.calc_crc(data, @crc32_defn)


```

### Performance

cerlc is slower than Erlang's built in crc32() function and CRC's implemented using 'C' language NIF's.

cerlc is useful for quickly generating CRC's with little code and no other dependencies.

### Sources of CRC Definitions
[Javascript CRC Calculator](http://www.sunshine2k.de/coding/javascript/crc/crc_js.html)

[crccalc](https://crccalc.com/)


CHANGELOG.md
# Changelog

## v0.2.1

* Add EDoc documentation generation


## v0.2.0

* Switch to configurable CRC function generator


## v0.1.0

* Initial release to Hex.pm

 

test/cerlc_tests.erl
-module(cerlc_tests).

-include_lib("eunit/include/eunit.hrl").

% Test each predefined CRC algorithm
crc8_test() -> crc_test_common(crc8, 16#F4).
crc8_sae_j1850_test() -> crc_test_common(crc8_sae_j1850, 16#4B).
crc8_sae_j1850_zero_test() -> crc_test_common(crc8_sae_j1850_zero, 16#37).
crc8_8h2f_test() -> crc_test_common(crc8_8h2f, 16#DF).
crc8_cdma2000_test() -> crc_test_common(crc8_cdma2000, 16#DA).
crc8_darc_test() -> crc_test_common(crc8_darc, 16#15).
crc8_dvb_s2_test() -> crc_test_common(crc8_dvb_s2, 16#BC).
crc8_ebu_test() -> crc_test_common(crc8_ebu, 16#97).
crc8_icode_test() -> crc_test_common(crc8_icode, 16#7E).
crc8_itu_test() -> crc_test_common(crc8_itu, 16#A1).
crc8_maxim_test() -> crc_test_common(crc8_maxim, 16#A1).
crc8_sensirion_test() -> crc_test_common(crc8_sensirion, 16#92, [16#BE, 16#EF]).
crc8_rohc_test() -> crc_test_common(crc8_rohc, 16#D0).
crc8_wcdma_test() -> crc_test_common(crc8_wcdma, 16#25).
crc16_ccitt_zero_test() -> crc_test_common(crc16_ccitt_zero, 16#31C3).
crc16_arc_test() -> crc_test_common(crc16_arc, 16#BB3D).
crc16_aug_ccitt_test() -> crc_test_common(crc16_aug_ccitt, 16#E5CC).
crc16_buypass_test() -> crc_test_common(crc16_buypass, 16#FEE8).
crc16_ccitt_false_test() -> crc_test_common(crc16_ccitt_false, 16#29B1).
crc16_cdma2000_test() -> crc_test_common(crc16_cdma2000, 16#4C06).
crc16_dds_110_test() -> crc_test_common(crc16_dds_110, 16#9ECF).
crc16_dect_r_test() -> crc_test_common(crc16_dect_r, 16#007E).
crc16_dect_x_test() -> crc_test_common(crc16_dect_x, 16#007F).
crc16_dnp_test() -> crc_test_common(crc16_dnp, 16#EA82).
crc16_en_13757_test() -> crc_test_common(crc16_en_13757, 16#C2B7).
crc16_genibus_test() -> crc_test_common(crc16_genibus, 16#D64E).
crc16_maxim_test() -> crc_test_common(crc16_maxim, 16#44C2).
crc16_mcrf4xx_test() -> crc_test_common(crc16_mcrf4xx, 16#6F91).
crc16_riello_test() -> crc_test_common(crc16_riello, 16#63D0).
crc16_t10_dif_test() -> crc_test_common(crc16_t10_dif, 16#D0DB).
crc16_teledisk_test() -> crc_test_common(crc16_teledisk, 16#0FB3).
crc16_tms37157_test() -> crc_test_common(crc16_tms37157, 16#26B1).
crc16_usb_test() -> crc_test_common(crc16_usb, 16#B4C8).
crc16_a_test() -> crc_test_common(crc16_a, 16#BF05).
crc16_kermit_test() -> crc_test_common(crc16_kermit, 16#2189).
crc16_modbus_test() -> crc_test_common(crc16_modbus, 16#4B37).
crc16_x_25_test() -> crc_test_common(crc16_x_25, 16#906E).
crc16_xmodem_test() -> crc_test_common(crc16_xmodem, 16#31C3).
crc32_test() -> crc_test_common(crc32, 16#CBF43926).
crc32_bzip2_test() -> crc_test_common(crc32_bzip2, 16#FC891918).
crc32_c_test() -> crc_test_common(crc32_c, 16#E3069283).
crc32_d_test() -> crc_test_common(crc32_d, 16#87315576).
crc32_mpeg2_test() -> crc_test_common(crc32_mpeg2, 16#0376E6E7).
crc32_posix_test() -> crc_test_common(crc32_posix, 16#765E7680).
crc32_q_test() -> crc_test_common(crc32_q, 16#3010BF7F).
crc32_jamcrc_test() -> crc_test_common(crc32_jamcrc, 16#340BC6D9).
crc32_xfer_test() -> crc_test_common(crc32_xfer, 16#BD0BE338).
crc64_ecma_182_test() -> crc_test_common(crc64_ecma_182, 16#6C40DF5F0B497347).
crc64_go_iso_test() -> crc_test_common(crc64_go_iso, 16#B90956C775A41001).
crc64_we_test() -> crc_test_common(crc64_we, 16#62EC59E3F1A4F00A).
crc64_xz_test() -> crc_test_common(crc64_xz, 16#995DC9BBDF1939FA).

% Test all bit widths, un-reflected and reflected, with made up parameters
crc8_custom_config_test() -> crc_test_common({8, 16#12, 16#34, 16#56, false}, 16#36).
crc8_r_custom_config_test() -> crc_test_common({8, 16#12, 16#34, 16#56, true}, 16#22).
crc16_custom_config_test() -> crc_test_common({16, 16#1234, 16#5678, 16#9ABC, false}, 16#420C).
crc16_r_custom_config_test() -> crc_test_common({16, 16#1234, 16#5678, 16#9ABC, true}, 16#AE4A).
crc32_custom_config_test() -> crc_test_common({32, 16#12345678, 16#9ABCDEF0, 16#12345678, false}, 16#D6FCC8B0).
crc32_r_custom_config_test() -> crc_test_common({32, 16#12345678, 16#9ABCDEF0, 16#12345678, true}, 16#1357BA22).
crc64_custom_config_test() -> crc_test_common({64, 16#123456789ABCDEF0, 16#123456789ABCDEF0, 16#123456789ABCDEF0, false}, 16#980704F6123B6210).
crc64_r_custom_config_test() -> crc_test_common({64, 16#123456789ABCDEF0, 16#123456789ABCDEF0, 16#123456789ABCDEF0, true}, 16#1B851603B23D082C).

% Test passing binary data into CRC calculation
crc8_sensirion_bin_test() -> crc_test_common(crc8_sensirion, 16#92, <<16#BE, 16#EF>>).


crc_test_common(CrcConfig, ExpectedCrc) ->
  crc_test_common(CrcConfig, ExpectedCrc, "123456789").

crc_test_common(CrcConfig, ExpectedCrc, TestData) ->
  CrcDefn = cerlc:init(CrcConfig),
  ?assertEqual(ExpectedCrc, cerlc:calc_crc(TestData, CrcDefn)).


src/cerlc.erl
%
% @doc A configurable cyclic redundancy check (CRC) generator
%
-module(cerlc).

% Invoke HiPE compilation when possible
% This will have no effect with JIT enabled (i.e. OTP-24 and later)
-compile(native).

-export([init/1, calc_crc/2, crc8_fun/3, crc16_32_64_fun/3, crc16_32_64_r_fun/3]).

-record(cerlc, {
  init_value :: non_neg_integer(),
  final_xor_value :: non_neg_integer(),
  crc_fun :: function(),
  table :: tuple(),
  shift :: non_neg_integer(),
  mask :: non_neg_integer()
}).

% @doc
% Generate a record containing the CRC calculation function
% and associated configuration values,
% for the given CRC algorithm definition
%

-spec init(atom() | tuple()) -> #cerlc{}.
init(CrcDef) when is_atom(CrcDef) ->
  init(name_to_config(CrcDef));

% Generate 8-bit CRC function for un-reflected case
init({8, Polynomial, InitValue, FinalXorValue, false}) ->

  #cerlc{
    init_value = InitValue,
    final_xor_value = FinalXorValue,
    % 8 bit CRC function is same for normal and reflected cases
    crc_fun = fun cerlc:crc8_fun/3,
    table = gen_table(8, Polynomial, false),
    shift = 0,
    mask = 0
  };

% Generate 8-bit CRC function for reflected case
init({8, Polynomial, InitValue, FinalXorValue, true}) ->

  #cerlc{
    init_value = reflect(8, InitValue),
    final_xor_value = FinalXorValue,
    % 8 bit CRC function is same for normal and reflected cases
    crc_fun = fun cerlc:crc8_fun/3,
    table = gen_table(8, Polynomial, true),
    shift = 0,
    mask = 0
  };

% Generate 16, 32, or 64-bit CRC function for un-reflected case
init({Bits, Polynomial, InitValue, FinalXorValue, false}) ->

  #cerlc{
    init_value = InitValue,
    final_xor_value = FinalXorValue,
    crc_fun = fun cerlc:crc16_32_64_fun/3,
    table = gen_table(Bits, Polynomial, false),
    shift = bits_to_shift(Bits),
    mask = bits_to_mask(Bits)
  };

% Generate 16, 32, or 64-bit CRC functions for reflected case
init({Bits, Polynomial, InitValue, FinalXorValue, true}) ->

  #cerlc{
    init_value = reflect(Bits, InitValue),
    final_xor_value = FinalXorValue,
    crc_fun = fun cerlc:crc16_32_64_r_fun/3,
    table = gen_table(Bits, Polynomial, true),
    shift = 0,
    mask = bits_to_mask(Bits)
  }.

% @private
crc8_fun([CurByte | Rem], CurCrc, Info) ->
    % Table tuple is indexed from 1, not 0
    Index = (CurCrc bxor CurByte) + 1,
    NextCrc = element(Index, Info#cerlc.table),
    crc8_fun(Rem, NextCrc, Info);

crc8_fun([], CurCrc, Info) -> CurCrc bxor Info#cerlc.final_xor_value.

% @private
crc16_32_64_fun([CurByte | Rem], CurCrc, Info) ->
  % Table tuple is indexed from 1, not 0
  Index = (((CurCrc bsr Info#cerlc.shift) bxor CurByte) band 16#FF) + 1,
  NextCrc = ((CurCrc bsl 8) bxor element(Index, Info#cerlc.table)) band Info#cerlc.mask,
  crc16_32_64_fun(Rem, NextCrc, Info);

crc16_32_64_fun([], CurCrc, Info) -> CurCrc bxor Info#cerlc.final_xor_value.

% @private
crc16_32_64_r_fun([CurByte | Rem], CurCrc, Info) ->
  % Table tuple is indexed from 1, not 0
  Index = ((CurCrc bxor CurByte) band 16#FF) + 1,
  NextCrc = ((CurCrc bsr 8) bxor element(Index, Info#cerlc.table)) band Info#cerlc.mask,
  crc16_32_64_r_fun(Rem, NextCrc, Info);

crc16_32_64_r_fun([], CurCrc, Info) -> CurCrc bxor Info#cerlc.final_xor_value.

% @doc
% Calculate the CRC of a binary or list of data bytes,
% using the CRC definition generated by init/1
%
-spec calc_crc(binary() | list(), #cerlc{}) -> non_neg_integer().
calc_crc(Data, Info) when is_binary(Data) ->
  calc_crc(binary_to_list(Data), Info);

calc_crc(Data, Info) when is_list(Data) -> 
  (Info#cerlc.crc_fun)(Data, Info#cerlc.init_value, Info).

% Generate CRC look-up table for un-reflected and reflected cases
gen_table(Bits, Polynomial, false) ->
  HiBit = bits_to_hibit(Bits),
  Mask = bits_to_mask(Bits),
  Shift = bits_to_shift(Bits),
  gen_table(HiBit, Mask, Shift, Polynomial, [], 0);

gen_table(Bits, Polynomial, true) ->
  PolynomialR = reflect(Bits, Polynomial),
  Mask = bits_to_mask(Bits),
  gen_table_r(Mask, PolynomialR, [], 0).

% Generate un-reflected CRC lookup table
gen_table(_HiBit, _Mask, _Shift, _Polynomial, Table, 256) ->
  list_to_tuple(lists:reverse(Table));

gen_table(HiBit, Mask, Shift, Polynomial, Table, Divident) ->
  CurrByte = curr_byte(HiBit, Polynomial, (Divident bsl Shift), 0),
  gen_table(HiBit, Mask, Shift, Polynomial, [(CurrByte band Mask) | Table], Divident + 1).

curr_byte(_HiBit, _Polynomial, CurrByte, 8) -> CurrByte;

curr_byte(HiBit, Polynomial, CurrByte, Index) ->
  NextByte = case (CurrByte band HiBit) of
    0 -> CurrByte bsl 1;
    _ -> (CurrByte bsl 1) bxor Polynomial
  end,
  curr_byte(HiBit, Polynomial, NextByte, Index + 1).

% Generate reflected CRC lookup table
gen_table_r(_Mask, _Polynomial, Table, 256) ->
  list_to_tuple(lists:reverse(Table));

gen_table_r(Mask, Polynomial, Table, Divident) ->
  CurrByte = curr_byte_r(Polynomial, Divident, 0),
  gen_table_r(Mask, Polynomial, [(CurrByte band Mask) | Table], Divident + 1).

curr_byte_r(_Polynomial, CurrByte, 8) -> CurrByte;

curr_byte_r(Polynomial, CurrByte, Index) ->
  NextByte = case (CurrByte band 1) of
    0 -> CurrByte bsr 1;
    1 -> (CurrByte bsr 1) bxor Polynomial
  end,
  curr_byte_r(Polynomial, NextByte, Index + 1).

% Convert CRC name to config parameters
name_to_config(crc8) -> {8, 16#07, 16#00, 16#00, false};
name_to_config(crc8_sae_j1850) -> {8, 16#1D, 16#FF, 16#FF, false};
name_to_config(crc8_sae_j1850_zero) -> {8, 16#1D, 16#00, 16#00, false};
name_to_config(crc8_8h2f) -> {8, 16#2F, 16#FF, 16#FF, false};
name_to_config(crc8_cdma2000) -> {8, 16#9B, 16#FF, 16#00, false};
name_to_config(crc8_darc) -> {8, 16#39, 16#00, 16#00, true};
name_to_config(crc8_dvb_s2) -> {8, 16#D5, 16#00, 16#00, false};
name_to_config(crc8_ebu) -> {8, 16#1D, 16#FF, 16#00, true};
name_to_config(crc8_icode) -> {8, 16#1D, 16#FD, 16#00, false};
name_to_config(crc8_itu) -> {8, 16#07, 16#00, 16#55, false};
name_to_config(crc8_maxim) -> {8, 16#31, 16#00, 16#00, true};
name_to_config(crc8_sensirion) -> {8, 16#31, 16#FF, 16#00, false};
name_to_config(crc8_rohc) -> {8, 16#07, 16#FF, 16#00, true};
name_to_config(crc8_wcdma) -> {8, 16#9B, 16#00, 16#00, true};
name_to_config(crc16_ccitt_zero) -> {16, 16#1021, 16#0000, 16#0000, false};
name_to_config(crc16_arc) -> {16, 16#8005, 16#0000, 16#0000, true};
name_to_config(crc16_aug_ccitt) -> {16, 16#1021, 16#1D0F, 16#0000, false};
name_to_config(crc16_buypass) -> {16, 16#8005, 16#0000, 16#0000, false};
name_to_config(crc16_ccitt_false) -> {16, 16#1021, 16#FFFF, 16#0000, false};
name_to_config(crc16_cdma2000) -> {16, 16#C867, 16#FFFF, 16#0000, false};
name_to_config(crc16_dds_110) -> {16, 16#8005, 16#800D, 16#0000, false};
name_to_config(crc16_dect_r) -> {16, 16#0589, 16#0000, 16#0001, false};
name_to_config(crc16_dect_x) -> {16, 16#0589, 16#0000, 16#0000, false};
name_to_config(crc16_dnp) -> {16, 16#3D65, 16#0000, 16#FFFF, true};
name_to_config(crc16_en_13757) -> {16, 16#3D65, 16#0000, 16#FFFF, false};
name_to_config(crc16_genibus) -> {16, 16#1021, 16#FFFF, 16#FFFF, false};
name_to_config(crc16_maxim) -> {16, 16#8005, 16#0000, 16#FFFF, true};
name_to_config(crc16_mcrf4xx) -> {16, 16#1021, 16#FFFF, 16#0000, true};
name_to_config(crc16_riello) -> {16, 16#1021, 16#B2AA, 16#0000, true};
name_to_config(crc16_t10_dif) -> {16, 16#8BB7, 16#0000, 16#0000, false};
name_to_config(crc16_teledisk) -> {16, 16#A097, 16#0000, 16#0000, false};
name_to_config(crc16_tms37157) -> {16, 16#1021, 16#89EC, 16#0000, true};
name_to_config(crc16_usb) -> {16, 16#8005, 16#FFFF, 16#FFFF, true};
name_to_config(crc16_a) -> {16, 16#1021, 16#C6C6, 16#0000, true};
name_to_config(crc16_kermit) -> {16, 16#1021, 16#0000, 16#0000, true};
name_to_config(crc16_modbus) -> {16, 16#8005, 16#FFFF, 16#0000, true};
name_to_config(crc16_x_25) -> {16, 16#1021, 16#FFFF, 16#FFFF, true};
name_to_config(crc16_xmodem) -> {16, 16#1021, 16#0000, 16#0000, false};
name_to_config(crc32) -> {32, 16#04C11DB7, 16#FFFFFFFF, 16#FFFFFFFF, true};
name_to_config(crc32_bzip2) -> {32, 16#04C11DB7, 16#FFFFFFFF, 16#FFFFFFFF, false};
name_to_config(crc32_c) -> {32, 16#1EDC6F41, 16#FFFFFFFF, 16#FFFFFFFF, true};
name_to_config(crc32_d) -> {32, 16#A833982B, 16#FFFFFFFF, 16#FFFFFFFF, true};
name_to_config(crc32_mpeg2) -> {32, 16#04C11DB7, 16#FFFFFFFF, 16#00000000, false};
name_to_config(crc32_posix) -> {32, 16#04C11DB7, 16#00000000, 16#FFFFFFFF, false};
name_to_config(crc32_q) -> {32, 16#814141AB, 16#00000000, 16#00000000, false};
name_to_config(crc32_jamcrc) -> {32, 16#04C11DB7, 16#FFFFFFFF, 16#00000000, true};
name_to_config(crc32_xfer) -> {32, 16#000000AF, 16#00000000, 16#00000000, false};
name_to_config(crc64_ecma_182) -> {64, 16#42F0E1EBA9EA3693, 16#0, 16#0, false};

name_to_config(crc64_go_iso) ->
  {64, 16#000000000000001B, 16#FFFFFFFFFFFFFFFF, 16#FFFFFFFFFFFFFFFF, true};

name_to_config(crc64_we) ->
  {64, 16#42F0E1EBA9EA3693, 16#FFFFFFFFFFFFFFFF, 16#FFFFFFFFFFFFFFFF, false};

name_to_config(crc64_xz) ->
  {64, 16#42F0E1EBA9EA3693, 16#FFFFFFFFFFFFFFFF, 16#FFFFFFFFFFFFFFFF, true}.

%
% Utility functions
%

reflect(Bits, Value) ->
  reflect(Bits, Value, 0, 0).

reflect(Bits, _Value, Result, Index) when Index == Bits -> Result;

reflect(Bits, Value, Result, Index) ->
  NextResult = case (Value band (1 bsl Index)) of
    0 -> Result;
    _ -> Result bor (1 bsl ((Bits - 1) - Index))
  end,
  reflect(Bits, Value, NextResult, Index + 1).

bits_to_mask(Bits) -> (1 bsl Bits) - 1.

bits_to_hibit(Bits) -> 1 bsl (Bits - 1).

bits_to_shift(Bits) -> Bits - 8.


src/crc16_usb.erl
%%
%% Implement CRC-16/USB
%%
%% Input reflected: yes
%% Result reflected: yes
%% Polynomial:, 0 x8005
%% Initial Value:, 0 xFFFF
%% Final Xor Value:, 0 xFFFF
%%

-module(crc16_usb).

-compile(native).

-compile({inline,[table/0]}).

-export([crc/1, crc/2]).

-type uint16() :: 0..16#FFFF.

-spec crc(binary() | [byte()]) -> uint16().
crc(Data) when is_binary(Data) ->
  crc(binary_to_list(Data));

crc(Data) when is_list(Data) ->
  crc(16#FFFF, Data).

-spec crc(uint16(), [byte()]) -> uint16().
crc(Crc, [X | Rem]) ->
  % table() tuple is indexed from 1, not 0
  Index = ((Crc bxor X) band 16#FF) + 1,
  NextCrc = ((Crc bsr 8) bxor element(Index, table())) band 16#FFFF,
  crc(NextCrc, Rem);

crc(Crc, []) -> (Crc bxor 16#FFFF).

% Use reflected lookup table
-spec table() -> tuple().
table() ->
{
  16#0000, 16#C0C1, 16#C181, 16#0140, 16#C301, 16#03C0, 16#0280, 16#C241,
  16#C601, 16#06C0, 16#0780, 16#C741, 16#0500, 16#C5C1, 16#C481, 16#0440,
  16#CC01, 16#0CC0, 16#0D80, 16#CD41, 16#0F00, 16#CFC1, 16#CE81, 16#0E40,
  16#0A00, 16#CAC1, 16#CB81, 16#0B40, 16#C901, 16#09C0, 16#0880, 16#C841,
  16#D801, 16#18C0, 16#1980, 16#D941, 16#1B00, 16#DBC1, 16#DA81, 16#1A40,
  16#1E00, 16#DEC1, 16#DF81, 16#1F40, 16#DD01, 16#1DC0, 16#1C80, 16#DC41,
  16#1400, 16#D4C1, 16#D581, 16#1540, 16#D701, 16#17C0, 16#1680, 16#D641,
  16#D201, 16#12C0, 16#1380, 16#D341, 16#1100, 16#D1C1, 16#D081, 16#1040,
  16#F001, 16#30C0, 16#3180, 16#F141, 16#3300, 16#F3C1, 16#F281, 16#3240,
  16#3600, 16#F6C1, 16#F781, 16#3740, 16#F501, 16#35C0, 16#3480, 16#F441,
  16#3C00, 16#FCC1, 16#FD81, 16#3D40, 16#FF01, 16#3FC0, 16#3E80, 16#FE41,
  16#FA01, 16#3AC0, 16#3B80, 16#FB41, 16#3900, 16#F9C1, 16#F881, 16#3840,
  16#2800, 16#E8C1, 16#E981, 16#2940, 16#EB01, 16#2BC0, 16#2A80, 16#EA41,
  16#EE01, 16#2EC0, 16#2F80, 16#EF41, 16#2D00, 16#EDC1, 16#EC81, 16#2C40,
  16#E401, 16#24C0, 16#2580, 16#E541, 16#2700, 16#E7C1, 16#E681, 16#2640,
  16#2200, 16#E2C1, 16#E381, 16#2340, 16#E101, 16#21C0, 16#2080, 16#E041,
  16#A001, 16#60C0, 16#6180, 16#A141, 16#6300, 16#A3C1, 16#A281, 16#6240,
  16#6600, 16#A6C1, 16#A781, 16#6740, 16#A501, 16#65C0, 16#6480, 16#A441,
  16#6C00, 16#ACC1, 16#AD81, 16#6D40, 16#AF01, 16#6FC0, 16#6E80, 16#AE41,
  16#AA01, 16#6AC0, 16#6B80, 16#AB41, 16#6900, 16#A9C1, 16#A881, 16#6840,
  16#7800, 16#B8C1, 16#B981, 16#7940, 16#BB01, 16#7BC0, 16#7A80, 16#BA41,
  16#BE01, 16#7EC0, 16#7F80, 16#BF41, 16#7D00, 16#BDC1, 16#BC81, 16#7C40,
  16#B401, 16#74C0, 16#7580, 16#B541, 16#7700, 16#B7C1, 16#B681, 16#7640,
  16#7200, 16#B2C1, 16#B381, 16#7340, 16#B101, 16#71C0, 16#7080, 16#B041,
  16#5000, 16#90C1, 16#9181, 16#5140, 16#9301, 16#53C0, 16#5280, 16#9241,
  16#9601, 16#56C0, 16#5780, 16#9741, 16#5500, 16#95C1, 16#9481, 16#5440,
  16#9C01, 16#5CC0, 16#5D80, 16#9D41, 16#5F00, 16#9FC1, 16#9E81, 16#5E40,
  16#5A00, 16#9AC1, 16#9B81, 16#5B40, 16#9901, 16#59C0, 16#5880, 16#9841,
  16#8801, 16#48C0, 16#4980, 16#8941, 16#4B00, 16#8BC1, 16#8A81, 16#4A40,
  16#4E00, 16#8EC1, 16#8F81, 16#4F40, 16#8D01, 16#4DC0, 16#4C80, 16#8C41,
  16#4400, 16#84C1, 16#8581, 16#4540, 16#8701, 16#47C0, 16#4680, 16#8641,
  16#8201, 16#42C0, 16#4380, 16#8341, 16#4100, 16#81C1, 16#8081, 16#4040
}.

%% ====================================================================
%% Tests
%% ====================================================================

-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").

empty_list_test() ->
  ?assert(crc("") == 0).

empty_bin_test() ->
  ?assert(crc(<<>>) == 0).

std_list_test() ->
  ?assert(crc("123456789") == 16#B4C8).

bin_with_crc_test() ->
  ?assert(crc(<<16#31,16#32,16#33,16#34,16#35,16#36,16#37,16#38,16#39,16#C8,16#B4>>) == crc(<<0,0>>)).

-endif.


src/crc8.erl
%%
%% Implement CRC-8
%%
%% Input reflected: no
%% Result reflected: no
%% Polynomial: 0x07
%% Initial Value: 0x0
%% Final Xor Value: 0x0
%%

-module(crc8).

-compile(native).

-compile({inline,[table/0]}).

-export([crc/1, crc/2]).

-spec crc(binary() | [byte()]) -> byte().
crc(Data) when is_binary(Data) ->
  crc(binary_to_list(Data));

crc(Data) when is_list(Data) ->
  crc(16#0, Data).

-spec crc(byte(), [byte()]) -> byte().
crc(Crc, [X | Rem]) ->
  % table() tuple is indexed from 1, not 0
  Index = (Crc bxor X) + 1,
  NextCrc = element(Index, table()),
  crc(NextCrc, Rem);

crc(Crc, []) -> Crc.

-spec table() -> tuple().
table() ->
{
  16#00, 16#07, 16#0E, 16#09, 16#1C, 16#1B, 16#12, 16#15,
  16#38, 16#3F, 16#36, 16#31, 16#24, 16#23, 16#2A, 16#2D,
  16#70, 16#77, 16#7E, 16#79, 16#6C, 16#6B, 16#62, 16#65,
  16#48, 16#4F, 16#46, 16#41, 16#54, 16#53, 16#5A, 16#5D,
  16#E0, 16#E7, 16#EE, 16#E9, 16#FC, 16#FB, 16#F2, 16#F5,
  16#D8, 16#DF, 16#D6, 16#D1, 16#C4, 16#C3, 16#CA, 16#CD,
  16#90, 16#97, 16#9E, 16#99, 16#8C, 16#8B, 16#82, 16#85,
  16#A8, 16#AF, 16#A6, 16#A1, 16#B4, 16#B3, 16#BA, 16#BD,
  16#C7, 16#C0, 16#C9, 16#CE, 16#DB, 16#DC, 16#D5, 16#D2,
  16#FF, 16#F8, 16#F1, 16#F6, 16#E3, 16#E4, 16#ED, 16#EA,
  16#B7, 16#B0, 16#B9, 16#BE, 16#AB, 16#AC, 16#A5, 16#A2,
  16#8F, 16#88, 16#81, 16#86, 16#93, 16#94, 16#9D, 16#9A,
  16#27, 16#20, 16#29, 16#2E, 16#3B, 16#3C, 16#35, 16#32,
  16#1F, 16#18, 16#11, 16#16, 16#03, 16#04, 16#0D, 16#0A,
  16#57, 16#50, 16#59, 16#5E, 16#4B, 16#4C, 16#45, 16#42,
  16#6F, 16#68, 16#61, 16#66, 16#73, 16#74, 16#7D, 16#7A,
  16#89, 16#8E, 16#87, 16#80, 16#95, 16#92, 16#9B, 16#9C,
  16#B1, 16#B6, 16#BF, 16#B8, 16#AD, 16#AA, 16#A3, 16#A4,
  16#F9, 16#FE, 16#F7, 16#F0, 16#E5, 16#E2, 16#EB, 16#EC,
  16#C1, 16#C6, 16#CF, 16#C8, 16#DD, 16#DA, 16#D3, 16#D4,
  16#69, 16#6E, 16#67, 16#60, 16#75, 16#72, 16#7B, 16#7C,
  16#51, 16#56, 16#5F, 16#58, 16#4D, 16#4A, 16#43, 16#44,
  16#19, 16#1E, 16#17, 16#10, 16#05, 16#02, 16#0B, 16#0C,
  16#21, 16#26, 16#2F, 16#28, 16#3D, 16#3A, 16#33, 16#34,
  16#4E, 16#49, 16#40, 16#47, 16#52, 16#55, 16#5C, 16#5B,
  16#76, 16#71, 16#78, 16#7F, 16#6A, 16#6D, 16#64, 16#63,
  16#3E, 16#39, 16#30, 16#37, 16#22, 16#25, 16#2C, 16#2B,
  16#06, 16#01, 16#08, 16#0F, 16#1A, 16#1D, 16#14, 16#13,
  16#AE, 16#A9, 16#A0, 16#A7, 16#B2, 16#B5, 16#BC, 16#BB,
  16#96, 16#91, 16#98, 16#9F, 16#8A, 16#8D, 16#84, 16#83,
  16#DE, 16#D9, 16#D0, 16#D7, 16#C2, 16#C5, 16#CC, 16#CB,
  16#E6, 16#E1, 16#E8, 16#EF, 16#FA, 16#FD, 16#F4, 16#F3
}.

%% ====================================================================
%% Tests
%% ====================================================================

-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").

empty_list_test() ->
  ?assert(crc("") == 0).

empty_bin_test() ->
  ?assert(crc(<<>>) == 0).

std_list_test() ->
  ?assert(crc("123456789") == 16#F4).

bin_with_crc_test() ->
  ?assert(crc(<<16#31,16#32,16#33,16#34,16#35,16#36,16#37,16#38,16#39,16#F4>>) == 0).

-endif.
